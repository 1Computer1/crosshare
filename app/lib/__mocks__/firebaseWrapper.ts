const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
