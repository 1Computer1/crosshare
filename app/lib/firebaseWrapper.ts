declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
