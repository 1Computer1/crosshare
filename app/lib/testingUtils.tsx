import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
