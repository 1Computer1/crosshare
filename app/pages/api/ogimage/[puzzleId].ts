import { NextApiRequest, NextApiResponse } from 'next';
import { createCanvas, loadImage, PNGStream } from 'canvas';
import { isRight } from 'fp-ts/lib/Either';
import { PathReporter } from 'io-ts/lib/PathReporter';

import { App } from '../../../lib/firebaseWrapper';
import { DBPuzzleV, DBPuzzleT } from '../../../lib/dbtypes';

function getPng(puzzle: DBPuzzleT): Promise<PNGStream> {
  console.log('Generating png for ' + puzzle.t);

  const canvas = createCanvas(1200, 630);
  const ctx = canvas.getContext('2d');

  ctx.fillStyle = 'white';
  ctx.fillRect(0, 0, canvas.width, canvas.height);

  // Puzzle outline
  ctx.strokeStyle = 'black';
  ctx.lineWidth = 3;
  ctx.lineJoin = 'round';
  ctx.strokeRect(300, 15, 600, 600);

  // Grid lines
  ctx.lineWidth = 2;
  const widthDivision = 600 / puzzle.w;
  const heightDivision = 600 / puzzle.h;
  ctx.beginPath();
  for (let i = 1; i < puzzle.w; i += 1) {
    ctx.moveTo(300 + i * widthDivision, 15);
    ctx.lineTo(300 + i * widthDivision, 615);
  }
  for (let i = 1; i < puzzle.h; i += 1) {
    ctx.moveTo(300, 15 + i * heightDivision);
    ctx.lineTo(900, 15 + i * heightDivision);
  }
  ctx.stroke();

  // Grid squares
  for (let i = 0; i < puzzle.g.length; i += 1) {
    if (puzzle.g[i] !== '.') {
      continue;
    }
    const col = i % puzzle.w;
    const row = Math.floor(i / puzzle.w);

    ctx.fillStyle = 'black';
    ctx.fillRect(300 + col * widthDivision, 15 + row * heightDivision, widthDivision, heightDivision);
  }

  // Center Logo
  ctx.beginPath();
  ctx.arc(600, 315, 150, 0, 2 * Math.PI);
  ctx.fillStyle = 'white';
  ctx.strokeStyle = 'black';
  ctx.lineWidth = 1;
  ctx.fill();
  ctx.stroke();
  ctx.beginPath();
  ctx.arc(600, 315, 145, 0, 2 * Math.PI);
  ctx.fillStyle = '#EB984E';
  ctx.fill();

  return loadImage('./public/logo.svg').then((img) => {
    ctx.drawImage(img, 500, 215);
    return canvas.createPNGStream();
  });
}

async function getPuzzle(puzzleId: string): Promise<DBPuzzleT | null> {
  const db = App.firestore();
  let dbres;
  try {
    dbres = await db.collection('c').doc(puzzleId).get();
  } catch {
    return null;
  }
  if (!dbres.exists) {
    return null;
  }
  const validationResult = DBPuzzleV.decode(dbres.data());
  if (isRight(validationResult)) {
    return validationResult.right;
  } else {
    console.error(PathReporter.report(validationResult).join(','));
    return null;
  }
}

export default async function ogImage(req: NextApiRequest, res: NextApiResponse) {
  const { puzzleId } = req.query;
  if (!puzzleId || Array.isArray(puzzleId)) {
    return res.status(404).json({ statusCode: 404, message: 'bad puzzle params' });
  }
  const puzzle = await getPuzzle(puzzleId);
  if (!puzzle) {
    return res.status(404).json({ statusCode: 404, message: 'failed to get puzzle' });
  }
  res.setHeader('Cache-Control', 'public, max-age=172800, s-maxage=172800');
  res.writeHead(200, { 'Content-Type': 'image/png' });
  const png = await getPng(puzzle);
  png.pipe(res);
}
