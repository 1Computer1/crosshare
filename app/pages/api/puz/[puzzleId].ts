import { NextApiRequest, NextApiResponse } from 'next';
import { exportFile } from '../../../lib/converter';
import { getPuzzle } from '../../../lib/puzzleCache';

export default async function puz(req: NextApiRequest, res: NextApiResponse) {
  const { puzzleId } = req.query;
  if (!puzzleId || Array.isArray(puzzleId)) {
    return res
      .status(404)
      .json({ statusCode: 404, message: 'bad puzzle params' });
  }
  const puzzle = await getPuzzle(puzzleId);
  if (!puzzle) {
    return res
      .status(404)
      .json({ statusCode: 404, message: 'failed to get puzzle' });
  }
  res.setHeader('X-Robots-Tag', 'noindex');
  res.setHeader('Cache-Control', 'public, max-age=172800, s-maxage=172800');
  res.setHeader(
    'Content-Disposition',
    'inline; filename="' + puzzle.t.replace(/[^\w ]/g, '') + '.puz"'
  );
  res.writeHead(200, { 'Content-Type': 'application/x-crossword' });
  res.end(Buffer.from(exportFile(puzzle)));
}
