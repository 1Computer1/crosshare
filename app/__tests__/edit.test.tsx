import * as firebaseTesting from '@firebase/testing';
import { setApp, setAdminApp } from '../lib/firebaseWrapper';
import type * as firebaseAdminType from 'firebase-admin';
import { getUser, render, getMockedPuzzle, fireEvent } from '../lib/testingUtils';
import { PuzzleLoader } from '../pages/crosswords/[puzzleId]/edit';

jest.mock('../lib/firebaseWrapper');
jest.mock('../lib/WordDB');

afterEach(() => {
  jest.clearAllMocks();
});

let randoApp: firebase.app.App, app: firebase.app.App, admin: firebase.app.App;

const PUZZLEID = 'foobar';

beforeAll(async () => {
  randoApp = firebaseTesting.initializeTestApp({
    projectId,
    auth: {
      uid: 'tom', admin: false, firebase: {
        sign_in_provider: 'google'
      }
    }
  }) as firebase.app.App;
  app = firebaseTesting.initializeTestApp({
    projectId,
    auth: {
      uid: 'mike', admin: false, firebase: {
        sign_in_provider: 'google'
      }
    }
  }) as firebase.app.App;
  admin = firebaseTesting.initializeAdminApp({ projectId }) as firebase.app.App;
  setAdminApp(admin as unknown as firebaseAdminType.app.App);
});

afterAll(async () => {
  await app.delete();
  await admin.delete();
  await randoApp.delete();
});

const mike = getUser('mike', false);
const rando = getUser('tom', false);
const projectId = 'edit-test';

test('cannot edit if not correct user', async () => {
  sessionStorage.clear();
  localStorage.clear();

  await firebaseTesting.clearFirestoreData({ projectId });
  await admin.firestore().collection('c').doc(PUZZLEID).set(getMockedPuzzle({ a: 'mike' }));

  setApp(randoApp);

  const r = render(
    <PuzzleLoader puzzleId={PUZZLEID} auth={{ user: rando, isAdmin: false }} />, { user: rando }
  );

  expect(await r.findByText('You do not have permission to view this page', { exact: false })).toBeInTheDocument();
});

test('basic edit', async () => {
  sessionStorage.clear();
  localStorage.clear();

  await firebaseTesting.clearFirestoreData({ projectId });
  await admin.firestore().collection('c').doc(PUZZLEID).set(getMockedPuzzle({ a: 'mike' }));

  setApp(app);

  const r = render(
    <PuzzleLoader puzzleId={PUZZLEID} auth={{ user: mike, isAdmin: false }} />, { user: mike }
  );

  expect(await r.findByText('changes may take up to an hour', { exact: false })).toBeInTheDocument();
  fireEvent.click(r.getByTitle('Edit ESSAY'));
  fireEvent.change(r.getByLabelText('ESSAY'), { target: { value: 'A new clue for essay' } });
  fireEvent.click(r.getByText('Save'));

  expect(await r.findByText('A new clue for essay')).toBeInTheDocument();
  await checkSnapshot(PUZZLEID);

  fireEvent.click(r.getByTitle('Edit SETSA'));
  fireEvent.change(r.getByLabelText('SETSA'), { target: { value: 'A new clue for wedding' } });
  fireEvent.click(r.getByText('Save'));
  expect(await r.findByText('A new clue for wedding')).toBeInTheDocument();
  await checkSnapshot(PUZZLEID);
});

test('edit title', async () => {
  sessionStorage.clear();
  localStorage.clear();

  await firebaseTesting.clearFirestoreData({ projectId });
  await admin.firestore().collection('c').doc(PUZZLEID).set(getMockedPuzzle({ a: 'mike' }));

  setApp(app);

  const r = render(
    <PuzzleLoader puzzleId={PUZZLEID} auth={{ user: mike, isAdmin: false }} />, { user: mike }
  );

  expect(await r.findByText('changes may take up to an hour', { exact: false })).toBeInTheDocument();
  fireEvent.click(r.getByTitle('Edit Title'));
  fireEvent.change(r.getByPlaceholderText('Enter Title'), { target: { value: 'Well we got a new title' } });
  fireEvent.click(r.getByText('Save'));

  expect(await r.findByText('Well we got a new title')).toBeInTheDocument();
  await checkSnapshot(PUZZLEID);
});

async function checkSnapshot(puzzleId: string) {
  const res = await admin.firestore().collection('c').doc(puzzleId).get();
  const resData = res.data();
  if (!resData) {
    throw new Error('botch');
  }
  delete resData['p'];
  delete resData['cs'][0]['p'];
  expect(resData).toMatchSnapshot();
}

test('edit constructor note', async () => {
  sessionStorage.clear();
  localStorage.clear();

  await firebaseTesting.clearFirestoreData({ projectId });
  await admin.firestore().collection('c').doc(PUZZLEID).set(getMockedPuzzle({ a: 'mike' }));

  setApp(app);

  const r = render(
    <PuzzleLoader puzzleId={PUZZLEID} auth={{ user: mike, isAdmin: false }} />, { user: mike }
  );

  expect(await r.findByText('changes may take up to an hour', { exact: false })).toBeInTheDocument();
  fireEvent.click(r.getByTitle('Add Constructor Note'));
  fireEvent.change(r.getByPlaceholderText('Enter Constructor Note'), { target: { value: 'Here is our note' } });
  fireEvent.click(r.getByText('Save'));

  expect(await r.findByText('Here is our note')).toBeInTheDocument();
  await checkSnapshot(PUZZLEID);

  fireEvent.click(r.getByTitle('Edit Constructor Note'));
  fireEvent.change(r.getByPlaceholderText('Enter Constructor Note'), { target: { value: 'Here is our new note' } });
  fireEvent.click(r.getByText('Save'));

  expect(await r.findByText('Here is our new note')).toBeInTheDocument();
  expect(r.queryByTitle('Add Constructor Note')).toBeNull();
  await checkSnapshot(PUZZLEID);

  fireEvent.click(r.getByTitle('Delete Constructor Note'));
  await checkSnapshot(PUZZLEID);
  expect(await r.findByTitle('Add Constructor Note')).toBeInTheDocument();
});

test('security rules for puzzle edits', async () => {
  await firebaseTesting.clearFirestoreData({ projectId });
  await admin.firestore().collection('c').doc(PUZZLEID).set(getMockedPuzzle({ a: 'mike' }));
  await admin.firestore().collection('c').doc('other').set(getMockedPuzzle({ a: 'other' }));

  // Succeeds updating title
  await firebaseTesting.assertSucceeds(
    app.firestore().collection('c').doc(PUZZLEID).update({
      t: 'Here is a new title',
    })
  );

  // Succeeds updating clues
  const downs = getMockedPuzzle()['dc'];
  downs[2] = 'My new clue here';
  await firebaseTesting.assertSucceeds(
    app.firestore().collection('c').doc(PUZZLEID).update({
      dc: downs,
    })
  );

  // Fails for other user
  await firebaseTesting.assertFails(
    app.firestore().collection('c').doc('other').update({
      t: 'Here is a new title',
    })
  );

  // Fails for clues wrong length
  await firebaseTesting.assertFails(
    app.firestore().collection('c').doc(PUZZLEID).update({
      dc: downs.slice(0, -1)
    })
  );

  // Fails for setting featured
  await firebaseTesting.assertFails(
    app.firestore().collection('c').doc(PUZZLEID).update({
      f: true
    })
  );

  // Fails for setting moderated
  await firebaseTesting.assertFails(
    app.firestore().collection('c').doc(PUZZLEID).update({
      m: true
    })
  );

  // Fails for modifying grid
  const grid = getMockedPuzzle()['g'];
  grid[2] = 'M';
  await firebaseTesting.assertFails(
    app.firestore().collection('c').doc(PUZZLEID).update({
      g: grid,
    })
  );
});
