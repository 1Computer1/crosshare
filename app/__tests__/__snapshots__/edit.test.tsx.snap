// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic edit 1`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "A new clue for essay",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`basic edit 2`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "A new clue for essay",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "A new clue for wedding",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit blog post 1`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "bp": "Here is our new blog post",
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit blog post 2`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "bp": "Here is our new post",
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit blog post 3`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit constructor note 1`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cn": "Here is our note",
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit constructor note 2`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cn": "Here is our new note",
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit constructor note 3`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Raises, as young",
  "w": 5,
}
`;

exports[`edit title 1`] = `
Object {
  "a": "mike",
  "ac": Array [
    " Cobbler's forms",
    "Absolutely perfect",
    "Spike Lee's \\"She's ___ Have It\\"",
    "English class assignment",
    "Raises, as young",
  ],
  "an": Array [
    1,
    6,
    7,
    8,
    9,
  ],
  "c": null,
  "cs": Array [
    Object {
      "a": "fSEwJorvqOMK5UhNMHa4mu48izl1",
      "c": "A couple of two-worders today which I don't love, but I hope you all got it anyway!",
      "ch": false,
      "i": "LwgoVx0BAskM4wVJyoLj",
      "n": "Mike D",
      "t": 36.009,
    },
  ],
  "dc": Array [
    "Hybrid whose father is a lion",
    "___ of reality (wake-up call)",
    "___ date (makes wedding plans)",
    "Middle Ages invader",
    "Has a great night at the comedy club",
  ],
  "dn": Array [
    1,
    2,
    3,
    4,
    5,
  ],
  "g": Array [
    "L",
    "A",
    "S",
    "T",
    "S",
    "I",
    "D",
    "E",
    "A",
    "L",
    "G",
    "O",
    "T",
    "T",
    "A",
    "E",
    "S",
    "S",
    "A",
    "Y",
    "R",
    "E",
    "A",
    "R",
    "S",
  ],
  "h": 5,
  "m": false,
  "n": "Mike D",
  "pv": true,
  "t": "Well we got a new title",
  "w": 5,
}
`;
